dashboard/modules/job/tests/test_http_job_server.py

```diff
@@ -639,5 +643,110 @@ def test_jobs_env_hook(job_sdk_client: JobSubmissionClient):
+    with patch(
+        "ray.dashboard.modules.job.job_head.DataOrganizer", MockDataOrganizer
+    ), patch(
+        "ray.dashboard.modules.job.job_head.JobAgentSubmissionClient",
+        MockJobAgentSubmissionClient,
+    ), patch(
+        "ray.dashboard.modules.job.job_head.dashboard_consts.CANDIDATE_AGENT_NUMBER", 2
+    ):
+        agent_1 = (
+            "node1",
+            dict(
+                ipAddress="1.1.1.1",
+                httpPort=1,
+                grpcPort=1,
+                httpAddress="1.1.1.1:1",
+            ),
+        )
+        agent_2 = (
+            "node2",
+            dict(
+                ipAddress="2.2.2.2",
+                httpPort=2,
+                grpcPort=2,
+                httpAddress="2.2.2.2:2",
+            ),
+        )
+        agent_3 = (
+            "node3",
+            dict(
+                ipAddress="3.3.3.3",
+                httpPort=3,
+                grpcPort=3,
+                httpAddress="3.3.3.3:3",
+            ),
+        )
+
+        MockDataOrganizer.agents["node1"] = dict(
+            ipAddress="1.1.1.1",
+            httpPort=1,
+            grpcPort=1,
+            httpAddress="1.1.1.1:1",
+        )
+        job_head = MockJobHead()
+        job_agent_client = await job_head.choose_agent()
+        assert job_agent_client._address == "http://1.1.1.1:1"
+
+        del MockDataOrganizer.agents["node1"]
+        with pytest.raises(asyncio.TimeoutError):
+            await asyncio.wait_for(job_head.choose_agent(), timeout=3)
+
+        MockDataOrganizer.agents[agent_1[0]] = agent_1[1]
+        MockDataOrganizer.agents[agent_2[0]] = agent_2[1]
+        MockDataOrganizer.agents[agent_3[0]] = agent_3[1]
+
+        addresses = set()
+        for address in range(2):
+            job_agent_client = await job_head.choose_agent()
+            addresses.add(job_agent_client._address)
+        assert addresses == {"http://1.1.1.1:1", "http://2.2.2.2:2"}
+        addresses = set()
+        for address in range(2):
+            job_agent_client = await job_head.choose_agent()
+            addresses.add(job_agent_client._address)
+        assert addresses == {"http://1.1.1.1:1", "http://2.2.2.2:2"}
+
+        del MockDataOrganizer.agents[agent_1[0]]
+        addresses = set()
+        for address in range(2):
+            job_agent_client = await job_head.choose_agent()
+            addresses.add(job_agent_client._address)
+        assert addresses == {"http://3.3.3.3:3", "http://2.2.2.2:2"}
+        addresses = set()
+        for address in range(2):
+            job_agent_client = await job_head.choose_agent()
+            addresses.add(job_agent_client._address)
+        assert addresses == {"http://3.3.3.3:3", "http://2.2.2.2:2"}
+
+        del MockDataOrganizer.agents[agent_2[0]]
+        for _ in range(2):
+            job_agent_client = await job_head.choose_agent()
+            assert "http://3.3.3.3:3" == job_agent_client._address
+
```

Comment:
I think we should try to avoid using mocks unless absolutely necessary.  Could we refactor `choose_agent()` so that it's testable without mocks, perhaps by passing in its dependencies as arguments to `choose_agent()`?

Reply:
We discuss the refactoring of choose_agent [here](https://github.com/ray-project/ray/pull/28599#discussion_r974560492
). When the implementation of choose_agent is completed, I will change the test again.

Reply:
@Catch-Bull Is it feasible to write `choose_agent` in such a way that we avoid using mocks in this test?

Reply:
@architkulkarni , sure, I remove those mocks, please take look at it.